jupyter-org-client.el

```diff
@@ -416,6 +416,357 @@ the `syntax-table' will be set to that of the REPL buffers."
+      (forward-line -3)
+      (end-of-line))))
+
+;;;###autoload
+(defun jupyter-org-move-src-block (&optional below)
+  "Move source block before of after another.
+
+If BELOW is non-nil, move the block down, otherwise move it up."
+  (interactive)
+  ;;skip all this if there are no previous or next source block
+  (when (condition-case nil
+            (save-excursion
+              (if below
+                  (org-babel-next-src-block)
+                (org-babel-previous-src-block)))
+          (error nil))
+    (let* ((src (org-element-context))
+           (results-start (org-babel-where-is-src-block-result))
+           (results-end
+            (when results-start
+              (save-excursion
+                (goto-char results-start)
+                (goto-char (org-babel-result-end))
+                ;; if line is empty, take that empty line by moving down
+                (when (looking-at-p "[[:space:]]*$")
+                  (forward-line 1))
+                (point)))))
+      ;; kill from the start of the source block to the end of the results
+      ;; ... if there are no results, to the end of the source block
+      (kill-region
+       (org-element-property :begin src)
+       (or results-end (org-element-property :end src))))
+    (if below
+        ;; if below, move past the next source block or its result
+        (let ((next-src-block-head (org-babel-where-is-src-block-head)))
+          (if next-src-block-head
+              (goto-char next-src-block-head)
+            (org-babel-next-src-block))
+          (let ((next-src-block (org-element-context))
+                (next-results-start (org-babel-where-is-src-block-result)))
+            (if next-results-start
+                (progn
+                  (goto-char next-results-start)
+                  (goto-char (org-babel-result-end))
+                  (when (looking-at-p "[[:space:]]*$")
+                    (forward-line 1)))
+              (goto-char (org-element-property :end next-src-block)))))
+      ;; else, move to the begining of the previous block
+      (org-babel-previous-src-block))
+    (save-excursion (org-yank)))) ; keep cursor where the yank takes place
+
+;;;###autoload
+(defun jupyter-org-clear-all-results ()
+  "Clear all results in the buffer."
+  (interactive)
+  (save-excursion
+    (goto-char (point-min))
+    (while (org-babel-next-src-block)
+      (org-babel-remove-result))))
+
+;;;###autoload
+(defun jupyter-org-restart-kernel-execute-block ()
+  "Restart kernel and execute block."
+  (interactive)
+  (let* ((params (car jupyter-org--src-block-cache))
+         (jupyter-current-client
+          (buffer-local-value 'jupyter-current-client
+                              (org-babel-jupyter-initiate-session
+                               (alist-get :session params) params))))
+    ;; FIXME: retarting the kernel does not work
+    (jupyter-repl-restart-kernel))
+  (org-babel-execute-src-block-maybe))
+
+(with-eval-after-load 'hydra
```

Comment:
I would prefer not to use `with-eval-after-load` since that is mainly meant for init files and not necessarily libraries.

I don't really know if there is a clean way to do this without `with-eval-after-load` so I'm thinking it may be best to make a new package that explicitly depends on `hydra` and then we point the user to load that package in their init files? This wouldn't be very convenient though.

Reply:
Maybe we can do something like

```emacs-lisp
(defun jupyter-org-hydra/body ()
  (interactive)
  (unless (require 'hydra nil t)
    (error "Install the `hydra' package."))
  ;; Unbind this function now since it will be re-bound to the actual hydra
  ;; below
  (fmakunbound 'jupyter-org-hydra/body)
  (eval `(defhydra ...)))
```

Reply:
Quite a nice idea.
I'll implement that.